nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2BrowseNodeGenerator.java

```diff
@@ -42,49 +38,51 @@
 public class P2BrowseNodeGenerator
     implements BrowseNodeGenerator
 {
-  private static final String REMOTE_URL_PREFIX_REGEX = "(" + HTTP_NXRM_PREFIX + "|" + HTTPS_NXRM_PREFIX + ")";
+  private final Set<String> knownSubDirectories;
+
+  @Inject
+  public P2BrowseNodeGenerator() {
+    Set<String> knownFirstSegments = new HashSet<>();
+    knownFirstSegments.add("binary");
+    knownFirstSegments.add("features");
+    knownFirstSegments.add("plugins");
+
+    knownSubDirectories = Collections.unmodifiableSet(knownFirstSegments);
+  }
 
   @Override
   public List<BrowsePaths> computeAssetPaths(final Asset asset, @Nullable final Component component) {
-    String remote = component == null ? getRemoteForMetadataAsset(asset.name()) : getRemoteForPackage(asset.name());
-    List<BrowsePaths> paths = Lists.newArrayList(computeComponentPaths(asset, component).iterator());
-    paths.addAll(BrowsePaths
-        .fromPaths(Lists.newArrayList(
-            Splitter.on(DIVIDER).omitEmptyStrings().split(getRelativePath(asset.name(), remote)).iterator()),
-            true));
-    return paths;
+    List<String> assetPaths = splitPath(asset.name());
+
+    List<BrowsePaths> browsePaths = computeComponentPath(assetPaths, component);
+
+    browsePaths.addAll(BrowsePaths.fromPaths(Collections.singletonList(assetPaths.get(assetPaths.size() - 1)), false));
+    return browsePaths;
   }
 
   @Override
   public List<BrowsePaths> computeComponentPaths(final Asset asset, @Nullable final Component component) {
+    return computeComponentPath(splitPath(asset.name()), component);
+  }
+
+  private List<BrowsePaths> computeComponentPath(final List<String> assetPath, @Nullable final Component component) {
     List<String> pathParts = new ArrayList<>();
-    String remote = component == null ? getRemoteForMetadataAsset(asset.name()) : getRemoteForPackage(asset.name());
-    if (!remote.isEmpty()) {
-      pathParts.add(getRemoteWithoutPrefix(remote));
+    if (assetPath.size() > 1) {
+      pathParts.add(assetPath.get(0));
     }
+
     if (component != null) {
-      pathParts.addAll(Lists.newArrayList(Splitter.on('.').omitEmptyStrings().split(component.name()).iterator()));
+      if (!knownSubDirectories.contains(assetPath.get(0))) {
+        pathParts.add(assetPath.get(1));
```

Comment:
Potential OutOfBoundException: 
Condition on line 70 implies `assetPath.size()` can be <= 1 (perhaps 1?), but line 76 has `assetPath.get(1)` which would require to have a minimum of two elements.

Reply:
>  Condition on line 70 implies `assetPath.size()` can be <= 1 (perhaps 1?), but line 76 has `assetPath.get(1)` which would require to have a minimum of two elements.

It comes from an understanding of the path structure. If `component` is non-null we know we have an artifact (not metdata), at that point we know there are at least 2 segments in path (`{hash/}?{plugins|features|binary}/filename`). If the first segment is not known (e.g. binary, plugins or features) then we know that it has 3 segments (`{hash}/{plugins|features|binary}/filename`).